src/core/net/SRP_CODER.md

```diff
@@ -0,0 +1,353 @@
+
+- A value of `1` indicates more segments follow.
+- A value of `0` indicates that this is the last segment.
+
+The remaining bits after the most significant bit provide the numerical `uint` bit value chunks and follow big-endian order.
+
+```
+    First segment (2-8 bits)          Next segments (8 bits)
+   +---+---+---+---+---+---+         +---+---+---+---+---+---+---+---+
+   | C |  int value        |         | C |        7-bit value        |
+   +---+---+---+---+---+---+         +---+---+---+---+---+---+---+---+
+```
+
+Similar integer coding schemes ([Variable-Length Quantity (VLQ)](https://en.wikipedia.org/wiki/Variable-length_quantity)) are used in other protocols.
+
+### Examples
+
+Assuming the first segment is one byte (8-bit) long:
+
+- Numbers less than (2^7) 128 are encoded as a single byte.
+- Numbers between 128 and (2^14 - 1) 16,383 are encoded as two bytes.
+- Numbers between 16,384 and (2^21 - 1) 2,097,151 are encoded as three bytes.
+
+Value `4460 = 0x1234` is encoded as two bytes: `[0xa4, 0x34]`
+
+```
+                             (0x1234)
+  +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+  | 0   0   0   1   0   0   1   0 | 0   0   1   1   0   1   0   0 |
+  +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+          /         7-bit            / \         7-bit             \
+         /                          /   \                           \
+        /                          /     \                           \
+   +---+---+---+---+---+---+---+---+  +---+---+---+---+---+---+---+---+
+   | 1 | 0   1   0   0   1   0   0 |  | 0 | 0   1   1   0   1   0   0 |
+   +---+---+---+---+---+---+---+---+  +---+---+---+---+---+---+---+---+
+              (0xa4)                             (0x34)
+```
+
+Assuming the first segment is 6-bit long:
+
+- Numbers less than (2^5) 32 are encoded in the first segment.
+- Numbers between 32 and (2^12 - 1) 4,095 use two segments (the first 6-bit segment and an additional byte).
+
+## DNS Name and Label Format
+
+### Label Dispatch
+
+```
+    0   1   2   3   4   5   6   7
+  +---+---+---+---+---+---+---+---+
+  | Type  ... | ...               |
+  +---+---+---+---+---+---+---+---+
+
+```
+
+- `00 <6-bit len>`: Standard label encoding, where the lower 6 bits indicate the label's length (number of characters). This is followed by the label characters, totaling `len` in number.
+- `01 <6-bit len>`: Similar to the above, but the label begins with an underscore `'_'`. The `len` value here doesn't include the underscore itself, which isn't explicitly encoded after the dispatch byte.
+- `10 <6-bit offset first seg>`: This references a previously encoded label within the message. The 6-bit value in the dispatch byte serves as the first segment of a compact `uint` representing the offset to that label. This offset must point to the label dispatch byte of the previous label.
```

Comment:
I was just doing some testing of my code and noticed that there is confusion about what "offset" is. Is it the offset to the beginning of the message, or to the beginning of the data? I think it makes sense for it to be the beginning of the data, to make offsets 12 smaller and hence potentially encode in fewer bytes. But this is never said explicitly, and the DNS meaning of offset is "offset from the beginning of the message." E.g., C0 0C points to the first name in the question section. I think for our purposes we want that to be 00, not 0C.

Reply:
The offset is relative to the start of the coded message. 

> I think it makes sense for it to be the beginning of the data, to make offsets 12 smaller and hence potentially encode in fewer bytes.

There is no DNS header (12 bytes) in the coded message. The "Header block" already compresses this information.

Reply:
BTW, I went to fix this in my code and found that in fact I'd already implemented it correctly. :)